src/Paprika/Data/SlottedArray.cs

```diff
@@ -906,78 +910,198 @@ private void MarkAsDeleted(int index)
-                GetHashRef(writeAt) = GetHashRef(i);
-
-                // Copy everything, just overwrite the address
-                ref var destinationSlot = ref GetSlotRef(writeAt);
-                destinationSlot.KeyPreamble = slot.KeyPreamble;
-                destinationSlot.ItemAddress = writtenTo;
-
-                writeAt++;
-            }
-
-            // Memoize to what is read to
-            readTo = addr;
+            VectorizedDefragment128();
         }
+        else
+        {
+            ThrowNoVectorSupport();
+        }
+    }
 
-        // Finalize by setting the header
-        _header.Low = (ushort)(alive * Slot.TotalSize);
-        _header.High = (ushort)(_data.Length - writtenTo);
-        _header.Deleted = 0;
+     private void VectorizedDefragment256()
+     {
+         var count = _header.Low / Slot.TotalSize;
+         var jump = DoubleVectorSize / sizeof(ushort);
+         var aligned = AlignToDoubleVectorSize(_header.Low) / sizeof(ushort);
+
+         // Start from the first slot vector
+         var writeTo = HashesPerVector;
+
+         var preambleMask = Slot.GetKeyPreambleMaskAsVector256();
+         var deletedMask = Slot.GetKeyPreambleDeleteAsVector256();
+
+         ref var d = ref Unsafe.As<byte, ushort>(ref MemoryMarshal.GetReference(_data));
+         var i = writeTo;
+
+         while (i < aligned)
+         {
+             if (i + jump < aligned)
+             {
+                 var slotData = Vector256.LoadUnsafe(ref d, (UIntPtr)i);
+
+                 // Ignore all the even entries from the vector as the preamble data is contained within
+                 // the first 2 bytes (ushort) out of the 4 bytes slot.
+                 slotData = Avx2.BlendVariable(slotData, Vector256<ushort>.Zero, Vector256.Create((ushort)0b_1010_1010_1010_1010));
+                 
+                 var preambleData = Vector256.BitwiseAnd(slotData, preambleMask);
+                 
+                 if (Vector256.EqualsAny(preambleData, deletedMask))
+                 {
+                     // Some slots are deleted in this batch, process individually
+                     for (var j = 0; j < SlotsPerVector; j++)
+                     {
+                         var currentSlot = Unsafe.Add(ref Unsafe.As<byte, Slot>(ref MemoryMarshal.GetReference(_data)), i + j);
+                         if (!currentSlot.IsDeleted)
+                         {
+                             if (i != writeTo)
+                             {
+                                 CopySlot(i, writeTo);
+                             }
+
+                             writeTo += Slot.TotalSize;
+                         }
+
+                         i += Slot.TotalSize;
+                     }
+                 }
+                 else
+                 {
+                     // No deleted slots in this batch, copy the whole batch
+                     if (i != writeTo)
+                     {
+                         CopyBatch(i, writeTo);
```

Comment:
Ah, so we do batch copy but only if we find the whole batch valid. If there's a single delete we'll go one by one then?